File: finger-detection/puzzle.py
"""
FingerQuest Puzzle Module
==========================
Simple drag-and-arrange puzzle using finger tracking.

Interaction: Point your INDEX FINGER at a piece to grab it.
Move your finger to drag it. Point at empty space to release.
If released near the target slot, it snaps into place.
"""

import random
import math
import cv2
import numpy as np

# ---------------------------------------------------------------------------
# Colors (BGR)
# ---------------------------------------------------------------------------
WHITE = (255, 255, 255)
GREEN = (0, 255, 136)
RED = (255, 51, 102)
CYAN = (255, 255, 0)
YELLOW = (0, 255, 255)
PURPLE = (180, 80, 255)
BLUE = (255, 150, 50)
ORANGE = (0, 165, 255)
DARK_SLOT = (40, 40, 55)
SLOT_BORDER = (80, 80, 100)


class PuzzlePiece:
    """A single draggable puzzle piece."""

    def __init__(self, piece_id: int, label: str, x: int, y: int, size: int,
                 image=None):
        self.id = piece_id
        self.label = label
        self.x = x
        self.y = y
        self.size = size
        self.target_x = x
        self.target_y = y
        self.grabbed = False
        self.placed = False
        self.image = image  # optional (size, size, 3) BGR crop

    def contains(self, px: int, py: int) -> bool:
        half = self.size // 2
        return (self.x - half <= px <= self.x + half and
                self.y - half <= py <= self.y + half)

    def draw(self, img, hovered=False):
        half = self.size // 2
        x1, y1 = self.x - half, self.y - half
        x2, y2 = self.x + half, self.y + half

        if self.placed:
            border_color = (0, 180, 60)
            border_thick = 3
        elif self.grabbed:
            border_color = YELLOW
            border_thick = 3
        elif hovered:
            border_color = (0, 220, 255)
            border_thick = 3
        else:
            border_color = PURPLE
            border_thick = 2

        if self.image is not None:
            self._blit_image(img, x1, y1)
            cv2.rectangle(img, (x1, y1), (x2, y2), border_color, border_thick)
            return

        if self.placed:
            fill_color = GREEN
        elif self.grabbed:
            fill_color = CYAN
        elif hovered:
            fill_color = ORANGE
        else:
            fill_color = BLUE

        # Draw filled rectangle
        cv2.rectangle(img, (x1, y1), (x2, y2), fill_color, -1)
        cv2.rectangle(img, (x1, y1), (x2, y2), border_color, border_thick)

        # Label
        font_scale = 0.6
        (tw, th), _ = cv2.getTextSize(
            self.label, cv2.FONT_HERSHEY_SIMPLEX, font_scale, 2)
        tx = self.x - tw // 2
        ty = self.y + th // 2
        cv2.putText(img, self.label, (tx, ty),
                    cv2.FONT_HERSHEY_SIMPLEX, font_scale, WHITE, 2, cv2.LINE_AA)

    def _blit_image(self, img, x1, y1):
        """Paste self.image onto img at (x1, y1), clipped to img bounds."""
        h, w = img.shape[:2]
        src = self.image
        sh, sw = src.shape[:2]

        dst_x1, dst_y1 = max(x1, 0), max(y1, 0)
        dst_x2, dst_y2 = min(x1 + sw, w), min(y1 + sh, h)
        if dst_x2 <= dst_x1 or dst_y2 <= dst_y1:
            return

        src_x1, src_y1 = dst_x1 - x1, dst_y1 - y1
        src_x2, src_y2 = src_x1 + (dst_x2 - dst_x1), src_y1 + (dst_y2 - dst_y1)
        img[dst_y1:dst_y2, dst_x1:dst_x2] = src[src_y1:src_y2, src_x1:src_x2]


class PuzzleBoard:
    """Manages puzzle pieces, grid, grab/drag/snap logic."""

    def __init__(self, cols: int, rows: int, piece_size: int,
                 grid_x: int, grid_y: int,
                 scatter_x1: int, scatter_y1: int,
                 scatter_x2: int, scatter_y2: int):
        self.cols = cols
        self.rows = rows
        self.piece_size = piece_size
        self.grid_x = grid_x
        self.grid_y = grid_y
        self.scatter_x1 = scatter_x1
        self.scatter_y1 = scatter_y1
        self.scatter_x2 = scatter_x2
        self.scatter_y2 = scatter_y2

        self.pieces: list[PuzzlePiece] = []
        self.grabbed_piece: PuzzlePiece | None = None
        self.hovered_piece: PuzzlePiece | None = None
        self.is_solved = False
        self.move_count = 0
        self.grab_cooldown = 0  # frames before auto-grab

        self._build_pieces()

    def _build_pieces(self):
        self.pieces.clear()
        pid = 0
        for row in range(self.rows):
            for col in range(self.cols):
                pid += 1
                cx = self.grid_x + col * self.piece_size + self.piece_size // 2
                cy = self.grid_y + row * self.piece_size + self.piece_size // 2
                piece = PuzzlePiece(pid, str(pid), cx, cy, self.piece_size)
                piece.target_x = cx
                piece.target_y = cy
                self.pieces.append(piece)

    def set_images(self, images: list):
        """Assign an image crop to each piece, matching build (row-major) order."""
        for piece, image in zip(self.pieces, images):
            piece.image = image

    def shuffle(self):
        self.is_solved = False
        self.move_count = 0
        self.grabbed_piece = None
        self.grab_cooldown = 0
        for piece in self.pieces:
            piece.grabbed = False
            piece.placed = False
            # Scatter with spacing to reduce overlap
            for _ in range(20):
                px = random.randint(self.scatter_x1, self.scatter_x2)
                py = random.randint(self.scatter_y1, self.scatter_y2)
                too_close = False
                for other in self.pieces:
                    if other is piece:
                        break
                    if abs(px - other.x) < self.piece_size and abs(py - other.y) < self.piece_size:
                        too_close = True
                        break
                if not too_close:
                    piece.x = px
                    piece.y = py
                    break
            else:
                piece.x = random.randint(self.scatter_x1, self.scatter_x2)
                piece.y = random.randint(self.scatter_y1, self.scatter_y2)

    def update(self, finger_x: int, finger_y: int, finger_visible: bool):
        """Call every frame with index finger position."""
        if not finger_visible or finger_x < 0 or finger_y < 0:
            self.hovered_piece = None
            if self.grabbed_piece:
                self._release_at(self.grabbed_piece.x, self.grabbed_piece.y)
            self.grab_cooldown = 0
            return

        # Find piece under finger
        found = None
        for piece in reversed(self.pieces):
            if piece.placed:
                continue
            if piece.contains(finger_x, finger_y):
                found = piece
                break

        self.hovered_piece = found

        if self.grabbed_piece:
            self.grabbed_piece.x = finger_x
            self.grabbed_piece.y = finger_y
            self.grab_cooldown = 0
        elif found:
            self.grab_cooldown += 1
            if self.grab_cooldown >= 10:  # ~0.3s
                self._grab(found)
                self.grab_cooldown = 0
        else:
            self.grab_cooldown = 0

    def _grab(self, piece: PuzzlePiece):
        piece.grabbed = True
        self.grabbed_piece = piece
        self.move_count += 1

    def release(self):
        if self.grabbed_piece is None:
            return
        self._release_at(self.grabbed_piece.x, self.grabbed_piece.y)

    def _release_at(self, px: int, py: int):
        if self.grabbed_piece is None:
            return
        piece = self.grabbed_piece
        piece.grabbed = False
        self.grabbed_piece = None

        snap_dist = self.piece_size * 0.7
        dist = math.sqrt((piece.x - piece.target_x) ** 2 +
                         (piece.y - piece.target_y) ** 2)
        if dist <= snap_dist:
            piece.x = piece.target_x
            piece.y = piece.target_y
            piece.placed = True

        self.is_solved = all(p.placed for p in self.pieces)

    def draw(self, img):
        h, w = img.shape[:2]

        # Target grid
        grid_w = self.cols * self.piece_size
        grid_h = self.rows * self.piece_size
        cv2.rectangle(img,
                      (self.grid_x, self.grid_y),
                      (self.grid_x + grid_w, self.grid_y + grid_h),
                      SLOT_BORDER, 2)

        # Slot numbers
        pid = 0
        for row in range(self.rows):
            for col in range(self.cols):
                pid += 1
                cx = self.grid_x + col * self.piece_size + self.piece_size // 2
                cy = self.grid_y + row * self.piece_size + self.piece_size // 2
                half = self.piece_size // 2
                cv2.rectangle(img,
                              (cx - half, cy - half),
                              (cx + half, cy + half),
                              DARK_SLOT, -1)
                cv2.rectangle(img,
                              (cx - half, cy - half),
                              (cx + half, cy + half),
                              SLOT_BORDER, 1)
                cv2.putText(img, str(pid), (cx - 10, cy + 8),
                            cv2.FONT_HERSHEY_SIMPLEX, 0.45, (100, 100, 120),
                            1, cv2.LINE_AA)

        # Draw pieces: normal → hovered → grabbed (z-order)
        normal = [p for p in self.pieces
                  if not p.grabbed and p is not self.hovered_piece]
        hover = [p for p in self.pieces
                 if p is self.hovered_piece and not p.grabbed]
        grab = [p for p in self.pieces if p.grabbed]

        for piece in normal:
            piece.draw(img)
        for piece in hover:
            piece.draw(img, hovered=True)
        for piece in grab:
            piece.draw(img)

    def draw_status(self, img):
        h, w = img.shape[:2]
        y = h - 30
        cv2.putText(img, f"Moves: {self.move_count}", (20, y),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, WHITE, 1, cv2.LINE_AA)

        if self.is_solved:
            tw, _ = cv2.getTextSize("PUZZLE SOLVED!", cv2.FONT_HERSHEY_SIMPLEX,
                                    1.5, 3)[0]
            cv2.putText(img, "PUZZLE SOLVED!", (w // 2 - tw // 2, h // 2),
                        cv2.FONT_HERSHEY_SIMPLEX, 1.5, GREEN, 3, cv2.LINE_AA)


def create_default_puzzle(frame_w: int, frame_h: int) -> PuzzleBoard:
    """Create a 3x3 puzzle positioned for the frame size."""
    cols, rows = 3, 3
    ps = 70
    grid_w = cols * ps
    grid_h = rows * ps
    gx = frame_w - grid_w - 30
    gy = (frame_h - grid_h) // 2
    board = PuzzleBoard(cols, rows, ps, gx, gy, 30, 30, gx - 40, frame_h - 60)
    board.shuffle()
    return board


def slice_image_grid(img, cols: int, rows: int, tile_size: int) -> list:
    """Slice img into cols x rows tiles (row-major), each resized to
    tile_size x tile_size."""
    h, w = img.shape[:2]
    cell_w, cell_h = w // cols, h // rows
    tiles = []
    for row in range(rows):
        for col in range(cols):
            tile = img[row * cell_h:(row + 1) * cell_h,
                       col * cell_w:(col + 1) * cell_w]
            tiles.append(cv2.resize(tile, (tile_size, tile_size)))
    return tiles


def create_photo_puzzle(frame_w: int, frame_h: int, cropped_img) -> PuzzleBoard:
    """Create a 3x3 puzzle whose pieces are tiles cut from cropped_img."""
    cols, rows = 3, 3
    ps = 70
    grid_w = cols * ps
    grid_h = rows * ps
    gx = frame_w - grid_w - 30
    gy = (frame_h - grid_h) // 2
    board = PuzzleBoard(cols, rows, ps, gx, gy, 30, 30, gx - 40, frame_h - 60)
    board.set_images(slice_image_grid(cropped_img, cols, rows, ps))
    board.shuffle()
    return board

File: web/src/game/puzzle.ts
import { COLOR, GRAB_DWELL_MS } from "./constants";
import type { FingerState } from "./types";

export class PuzzlePiece {
  id: number;
  label: string;
  x: number;
  y: number;
  size: number;
  targetX: number;
  targetY: number;
  grabbed = false;
  placed = false;
  image: HTMLCanvasElement | null = null;

  constructor(id: number, label: string, x: number, y: number, size: number) {
    this.id = id;
    this.label = label;
    this.x = x;
    this.y = y;
    this.size = size;
    this.targetX = x;
    this.targetY = y;
  }

  contains(px: number, py: number): boolean {
    const half = this.size / 2;
    return (
      px >= this.x - half &&
      px <= this.x + half &&
      py >= this.y - half &&
      py <= this.y + half
    );
  }

  draw(ctx: CanvasRenderingContext2D, hovered = false) {
    const half = this.size / 2;
    const x1 = this.x - half;
    const y1 = this.y - half;

    let border: string = COLOR.purple;
    let borderW = 2;
    if (this.placed) {
      border = "#00b43c";
      borderW = 3;
    } else if (this.grabbed) {
      border = COLOR.yellow;
      borderW = 3;
    } else if (hovered) {
      border = "#00dcff";
      borderW = 3;
    }

    if (this.image) {
      ctx.drawImage(this.image, x1, y1, this.size, this.size);
      ctx.strokeStyle = border;
      ctx.lineWidth = borderW;
      ctx.strokeRect(x1, y1, this.size, this.size);
      return;
    }

    let fill: string = COLOR.blue;
    if (this.placed) fill = COLOR.green;
    else if (this.grabbed) fill = COLOR.cyan;
    else if (hovered) fill = COLOR.orange;

    ctx.fillStyle = fill;
    ctx.fillRect(x1, y1, this.size, this.size);
    ctx.strokeStyle = border;
    ctx.lineWidth = borderW;
    ctx.strokeRect(x1, y1, this.size, this.size);

    ctx.fillStyle = COLOR.white;
    ctx.font = "bold 18px sans-serif";
    ctx.textAlign = "center";
    ctx.fillText(this.label, this.x, this.y + 6);
    ctx.textAlign = "left";
  }
}

export class PuzzleBoard {
  cols: number;
  rows: number;
  pieceSize: number;
  gridX: number;
  gridY: number;
  scatterX1: number;
  scatterY1: number;
  scatterX2: number;
  scatterY2: number;

  pieces: PuzzlePiece[] = [];
  grabbedPiece: PuzzlePiece | null = null;
  hoveredPiece: PuzzlePiece | null = null;
  isSolved = false;
  moveCount = 0;
  private grabDwellMs = 0;

  constructor(
    cols: number,
    rows: number,
    pieceSize: number,
    gridX: number,
    gridY: number,
    scatterX1: number,
    scatterY1: number,
    scatterX2: number,
    scatterY2: number
  ) {
    this.cols = cols;
    this.rows = rows;
    this.pieceSize = pieceSize;
    this.gridX = gridX;
    this.gridY = gridY;
    this.scatterX1 = scatterX1;
    this.scatterY1 = scatterY1;
    this.scatterX2 = scatterX2;
    this.scatterY2 = scatterY2;
    this.buildPieces();
  }

  private buildPieces() {
    this.pieces = [];
    let pid = 0;
    for (let row = 0; row < this.rows; row++) {
      for (let col = 0; col < this.cols; col++) {
        pid++;
        const cx = this.gridX + col * this.pieceSize + this.pieceSize / 2;
        const cy = this.gridY + row * this.pieceSize + this.pieceSize / 2;
        const piece = new PuzzlePiece(pid, String(pid), cx, cy, this.pieceSize);
        piece.targetX = cx;
        piece.targetY = cy;
        this.pieces.push(piece);
      }
    }
  }

  setImages(images: HTMLCanvasElement[]) {
    this.pieces.forEach((piece, i) => {
      if (images[i]) piece.image = images[i];
    });
  }

  shuffle() {
    this.isSolved = false;
    this.moveCount = 0;
    this.grabbedPiece = null;
    this.grabDwellMs = 0;

    for (const piece of this.pieces) {
      piece.grabbed = false;
      piece.placed = false;
      let placedOk = false;
      for (let attempt = 0; attempt < 20; attempt++) {
        const px = this.scatterX1 + Math.random() * (this.scatterX2 - this.scatterX1);
        const py = this.scatterY1 + Math.random() * (this.scatterY2 - this.scatterY1);
        let tooClose = false;
        for (const other of this.pieces) {
          if (other === piece) break;
          if (
            Math.abs(px - other.x) < this.pieceSize &&
            Math.abs(py - other.y) < this.pieceSize
          ) {
            tooClose = true;
            break;
          }
        }
        if (!tooClose) {
          piece.x = px;
          piece.y = py;
          placedOk = true;
          break;
        }
      }
      if (!placedOk) {
        piece.x = this.scatterX1 + Math.random() * (this.scatterX2 - this.scatterX1);
        piece.y = this.scatterY1 + Math.random() * (this.scatterY2 - this.scatterY1);
      }
    }
  }

  update(finger: FingerState, dtMs: number) {
    if (!finger.visible) {
      this.hoveredPiece = null;
      if (this.grabbedPiece) this.releaseAt();
      this.grabDwellMs = 0;
      return;
    }

    let found: PuzzlePiece | null = null;
    for (let i = this.pieces.length - 1; i >= 0; i--) {
      const piece = this.pieces[i];
      if (piece.placed) continue;
      if (piece.contains(finger.x, finger.y)) {
        found = piece;
        break;
      }
    }
    this.hoveredPiece = found;

    if (this.grabbedPiece) {
      this.grabbedPiece.x = finger.x;
      this.grabbedPiece.y = finger.y;
      this.grabDwellMs = 0;
    } else if (found) {
      this.grabDwellMs += dtMs;
      if (this.grabDwellMs >= GRAB_DWELL_MS) {
        this.grab(found);
        this.grabDwellMs = 0;
      }
    } else {
      this.grabDwellMs = 0;
    }
  }

  private grab(piece: PuzzlePiece) {
    piece.grabbed = true;
    this.grabbedPiece = piece;
    this.moveCount++;
  }

  release() {
    this.releaseAt();
  }

  private releaseAt() {
    if (!this.grabbedPiece) return;
    const piece = this.grabbedPiece;
    piece.grabbed = false;
    this.grabbedPiece = null;

    const snapDist = this.pieceSize * 0.7;
    const dist = Math.hypot(piece.x - piece.targetX, piece.y - piece.targetY);
    if (dist <= snapDist) {
      piece.x = piece.targetX;
      piece.y = piece.targetY;
      piece.placed = true;
    }
    this.isSolved = this.pieces.every((p) => p.placed);
  }

  draw(ctx: CanvasRenderingContext2D) {
    const gridW = this.cols * this.pieceSize;
    const gridH = this.rows * this.pieceSize;
    ctx.strokeStyle = COLOR.slotBorder;
    ctx.lineWidth = 2;
    ctx.strokeRect(this.gridX, this.gridY, gridW, gridH);

    let pid = 0;
    for (let row = 0; row < this.rows; row++) {
      for (let col = 0; col < this.cols; col++) {
        pid++;
        const cx = this.gridX + col * this.pieceSize + this.pieceSize / 2;
        const cy = this.gridY + row * this.pieceSize + this.pieceSize / 2;
        const half = this.pieceSize / 2;
        ctx.fillStyle = COLOR.slotBg;
        ctx.fillRect(cx - half, cy - half, this.pieceSize, this.pieceSize);
        ctx.strokeStyle = COLOR.slotBorder;
        ctx.lineWidth = 1;
        ctx.strokeRect(cx - half, cy - half, this.pieceSize, this.pieceSize);
        ctx.fillStyle = "#64647a";
        ctx.font = "12px sans-serif";
        ctx.textAlign = "center";
        ctx.fillText(String(pid), cx, cy + 4);
        ctx.textAlign = "left";
      }
    }

    const normal = this.pieces.filter((p) => !p.grabbed && p !== this.hoveredPiece);
    const hover = this.pieces.filter((p) => p === this.hoveredPiece && !p.grabbed);
    const grab = this.pieces.filter((p) => p.grabbed);

    for (const p of normal) p.draw(ctx);
    for (const p of hover) p.draw(ctx, true);
    for (const p of grab) p.draw(ctx);
  }

  drawStatus(ctx: CanvasRenderingContext2D, frameW: number, frameH: number) {
    ctx.fillStyle = COLOR.white;
    ctx.font = "13px sans-serif";
    ctx.textAlign = "left";
    ctx.fillText(`Moves: ${this.moveCount}`, 20, frameH - 20);

    if (this.isSolved) {
      ctx.fillStyle = COLOR.green;
      ctx.font = "bold 32px sans-serif";
      ctx.textAlign = "center";
      ctx.fillText("PUZZLE SOLVED!", frameW / 2, frameH / 2);
      ctx.textAlign = "left";
    }
  }
}

export function createDefaultPuzzle(frameW: number, frameH: number): PuzzleBoard {
  const cols = 3;
  const rows = 3;
  const ps = 70;
  const gridW = cols * ps;
  const gridH = rows * ps;
  const gx = frameW - gridW - 30;
  const gy = (frameH - gridH) / 2;
  const board = new PuzzleBoard(cols, rows, ps, gx, gy, 30, 30, gx - 40, frameH - 60);
  board.shuffle();
  return board;
}

/** Slice a region of sourceCanvas into cols x rows tiles, each tileSize x tileSize. */
export function sliceImageGrid(
  source: HTMLCanvasElement,
  rectX: number,
  rectY: number,
  rectW: number,
  rectH: number,
  cols: number,
  rows: number,
  tileSize: number
): HTMLCanvasElement[] {
  const cellW = rectW / cols;
  const cellH = rectH / rows;
  const tiles: HTMLCanvasElement[] = [];

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      const tile = document.createElement("canvas");
      tile.width = tileSize;
      tile.height = tileSize;
      const ctx = tile.getContext("2d")!;
      ctx.drawImage(
        source,
        rectX + col * cellW,
        rectY + row * cellH,
        cellW,
        cellH,
        0,
        0,
        tileSize,
        tileSize
      );
      tiles.push(tile);
    }
  }
  return tiles;
}

export function createPhotoPuzzle(
  frameW: number,
  frameH: number,
  source: HTMLCanvasElement,
  rectX: number,
  rectY: number,
  rectW: number,
  rectH: number
): PuzzleBoard {
  const cols = 3;
  const rows = 3;
  const ps = 70;
  const gridW = cols * ps;
  const gridH = rows * ps;
  const gx = frameW - gridW - 30;
  const gy = (frameH - gridH) / 2;
  const board = new PuzzleBoard(cols, rows, ps, gx, gy, 30, 30, gx - 40, frameH - 60);
  board.setImages(sliceImageGrid(source, rectX, rectY, rectW, rectH, cols, rows, ps));
  board.shuffle();
  return board;
}
